### PR TITLE
feat(python,rust): consolidate & improve all casting failure error messages

### PR DESCRIPTION
**Pull request by JulianCologne**
This PR consolidates (unifies) and improves all different error messages the user sees when doing different kinds of casting operations. (`str.strptime`/`str.to_date`, `cast`, and casting in `read_csv` all have different messages).

There is now **1 single function** that handles the casting error messages for all those cases.

Examples

```python
df = pl.DataFrame({
    'x': ['2020-01-01', 'hello world', '2022-01-01', 'ups', 'hello world'],
})

DATA = '''x
0
255
256
300
300'''
```

## `str.to_date` & `str.strptime` (message 1)

```python
df.with_columns(
    # date=pl.col("x").str.to_date()
    date=pl.col("x").str.strptime(pl.Date)
)
strict date parsing failed for 3 value(s) (2 unique): ["hello world", "ups"]

You might want to try:
- setting `strict=False`
- setting `exact=False` (note: this is much slower!)
- explicitly specifying `format`
```

## `cast` (message 2)

```python
df.with_columns(
    date=pl.col("x").cast(pl.Date),
)
# or
df.cast(pl.Date)
strict conversion from `str` to `date` failed for column: x, value(s) ["ups", "hello world"]; if you were trying to cast Utf8 to temporal dtypes, consider using `strptime`
```

## `read_csv` (message 3)

```python
pl.read_csv(
    source=StringIO(DATA),
    dtypes={'x': pl.UInt8},
)
parsing to `u8` failed for column: x, value(s) [256, 300, 1000];
```

# New Format (1 single format for all)

With the refactoring there is a single function that handles all those messages and can still be adjusted for different types

### New Message (date casting)

```python
Conversion from `str` to `date` failed in column 'x' for 3 of 5 values (2 unique failures): ["hello world", "ups"]

You might want to try:
- setting `strict=False
- providing a format or checking the format string
- setting `exact=False` (note: this is much slower!)
```

### New Message (u8 casting)

```python
Conversion from `i32` to `u8` failed in column 'x' for 3 of 5 values (2 unique failures): [256, 300]
```

---

Just saw there were some changes during my refactoring which removed the unique count to reduce compute in error message. This could easily be adjusted now that there is only one function to handle all those error messages.

However, I am personally against removing the unique count!
I am using polars for over a year in different companies for production transformation and etl pipelines. I vastly prefer polars taking some extra 10,20, 50 ms (even seconds lol) to produce the best error message it can for me to fix the problem. I don't care at all if polars saves me 10ms but I have a harder time fixing the bug. 
Still, even without the unique count, I think this error message is still very good! =)

### CODE REVIEW

**Review by MarcoGorelli — COMMENTED** [review]
Nice, generally on-board, thanks for having done this again!

Could you resolve the merge conflicts and remove the extra computation please?

**Comment by MarcoGorelli** [comment]
Is this right?
```python
In [5]: df.with_columns(
   ...:     date=pl.col("x").cast(pl.Date),
   ...: )
---------------------------------------------------------------------------
ComputeError                              Traceback (most recent call last)
Cell In[5], line 1
----> 1 df.with_columns(
      2     date=pl.col("x").cast(pl.Date),
      3 )

File ~/polars-dev/py-polars/polars/dataframe/frame.py:7944, in DataFrame.with_columns(self, *exprs, **named_exprs)
   7796 def with_columns(
   7797     self,
   7798     *exprs: IntoExpr | Iterable[IntoExpr],
   7799     **named_exprs: IntoExpr,
   7800 ) -> DataFrame:
   7801     """
   7802     Add columns to this DataFrame.
   7803
   (...)
   7942
   7943     """
-> 7944     return self.lazy().with_columns(*exprs, **named_exprs).collect(_eager=True)

File ~/polars-dev/py-polars/polars/utils/deprecation.py:100, in deprecate_renamed_parameter.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
     95 @wraps(function)
     96 def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
     97     _rename_keyword_argument(
     98         old_name, new_name, kwargs, function.__name__, version
     99     )
--> 100     return function(*args, **kwargs)

File ~/polars-dev/py-polars/polars/lazyframe/frame.py:1787, in LazyFrame.collect(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, slice_pushdown, comm_subplan_elim, comm_subexpr_elim, no_optimization, streaming, _eager)
   1774     comm_subplan_elim = False
   1776 ldf = self._ldf.optimization_toggle(
   1777     type_coercion,
   1778     predicate_pushdown,
   (...)
   1785     _eager,
   1786 )
-> 1787 return wrap_df(ldf.collect())

ComputeError: Conversion from `str` to `date` failed in column 'x' for 3 of 5 values (first few failures: ["hello world", "ups", "hello world"])

You might want to try:
- setting `strict=False
- providing a format or checking the format string
- setting `exact=False` (note: this is much slower!)
```
`.cast` doesn't have a `strict` argument, nor a format one, nor `exact`

**Comment by JulianCologne** [comment]
Good point! I will adjust the message text regarding "format" and "exact".

"strict" is available for all the different implementations though, isn't it?

**Comment by MarcoGorelli** [comment]
sorry yes, strict is fine, just the rest of the message needs adjusting

**Comment by JulianCologne** [comment]
I added references to `str.strptime`, `str.to_date` and `str.to_datetime` and the format string when converting dates/datetimes.

I also intentional removed the `exact` reference because it is easily discoverable by the user but it has a big performance footprint, rather limited use cases and there is not even an example of what it does in the docs so it might even confuse users.

File: crates/polars-core/src/utils/series.rs [hunk 1]
```diff
@@ -40,7 +40,30 @@ pub fn ensure_sorted_arg(s: &Series, operation: &str) -> PolarsResult<()> {
...
-    input.filter_threaded(&failure_mask, false)?.unique()
+    let failures = input.filter_threaded(&failure_mask, false)?;
+    let first_failures = failures.slice(0, 3);
+    let n_failures = failures.len();
+
+    let additional_info = match (input.dtype(), output.dtype()) {
+        (DataType::Utf8, DataType::Date | DataType::Datetime(_, _)) => {
+            "\n\nYou might want to try:\n\
+            - setting `strict=False` to set values to `null` that cannot be converted\n\
+            - using `str.strptime`, `str.to_date` or `str.to_datetime` and provide a format string"
```
Comment (MarcoGorelli) [inline crates/polars-core/src/utils/series.rs:53]:
```suggestion
            - using `str.strptime`, `str.to_date`, or `str.to_datetime` and providing a format string"
```

File: crates/polars-core/src/utils/series.rs [hunk 2]
```diff
@@ -40,7 +40,30 @@ pub fn ensure_sorted_arg(s: &Series, operation: &str) -> PolarsResult<()> {
...
     let failure_mask = !input.is_null() & output.is_null();
-    input.filter_threaded(&failure_mask, false)?.unique()
+    let failures = input.filter_threaded(&failure_mask, false)?;
+    let first_failures = failures.slice(0, 3);
+    let n_failures = failures.len();
+
+    let additional_info = match (input.dtype(), output.dtype()) {
+        (DataType::Utf8, DataType::Date | DataType::Datetime(_, _)) => {
+            "\n\nYou might want to try:\n\
+            - setting `strict=False` to set values to `null` that cannot be converted\n\
```
Comment (MarcoGorelli) [inline crates/polars-core/src/utils/series.rs:52]:
```suggestion
            - setting `strict=False` to set values that cannot be converted to `null`\n\
```

File: crates/polars-core/src/utils/series.rs [hunk 3]
```diff
@@ -40,7 +40,30 @@ pub fn ensure_sorted_arg(s: &Series, operation: &str) -> PolarsResult<()> {
...
+            - using `str.strptime`, `str.to_date` or `str.to_datetime` and provide a format string"
+        },
+        _ => "",
+    };
+
+    polars_bail!(
+        ComputeError:
+        "Conversion from `{}` to `{}` failed in column '{}' for {} of {} values (first few failures: {}){}",
```
Comment (MarcoGorelli) [inline crates/polars-core/src/utils/series.rs:60]:
```suggestion
        "Conversion from `{}` to `{}` failed in column '{}' for {} out of {} values (first few failures: {}){}",
```

**Review by MarcoGorelli — COMMENTED** [review]
Cool, so now we get
```python
In [19]: df.with_columns(
    ...:     date=pl.col("x").cast(pl.Date),
    ...: )
---------------------------------------------------------------------------

ComputeError: Conversion from `str` to `date` failed in column 'x' for 3 of 5 values (first few failures: ["hello world", "ups", "hello world"])

You might want to try:
- setting `strict=False` to set values to `null` that cannot be converted
- using `str.strptime`, `str.to_date` or `str.to_datetime` and provide a format string

In [20]: pl.read_csv(
    ...:     source=StringIO(DATA),
    ...:     dtypes={'x': pl.UInt8},
    ...: )
---------------------------------------------------------------------------

ComputeError: Conversion from `i32` to `u8` failed in column 'x' for 3 of 5 values (first few failures: [256, 300, 300])
```

Looks like an improvement to me! And also shares more code

Just got some really minor comments on the wording

**Comment by JulianCologne** [comment]
thanks for the comments.

I also condensed the code even more by removing the "failure-slice"and instead using `failures.fmt_list()` directly in the output. This has 2 benefits
- if you have multiple errors, you will also be able to see the last one (nice, because often you have 100x the same error in the beginning and you only see that one)
- you can change how much you see because this is controlled by `pl.Config.set_fmt_table_cell_list_len(n=10)`

```python
# 'x': ['2020-01-01', 'hello world', '2022-01-01', 'ups', 'hello world', 'x']

# >
Conversion from `str` to `date` failed in column 'x' for 4 out of 6 values: ["hello world", "ups", … "x"]

You might want to try:
- setting `strict=False` to set values that cannot be converted to `null`
- using `str.strptime`, `str.to_date`, or `str.to_datetime` and providing a format string
```

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me, thanks @JulianCologne !

**Pull request merged**